File: examples/kotlin/src/main/kotlin/adr/spine/boundary/InMemory.kt
// ── spine/boundary/in-memory — the deterministic adapters ──────────────────
// Everything that would otherwise be ambient — the clock, the id source, the log,
// the effect tap — lives here, behind the ports, so the whole system can be driven
// offline and re-driven identically.

package adr.spine.boundary

import adr.contract.Effect
import adr.spine.ports.Bus
import adr.spine.ports.Clock
import adr.spine.ports.IdSource
import adr.spine.ports.Sink
import adr.spine.pure.CommandId
import adr.spine.pure.EffectKey
import adr.spine.pure.KeyedEffect
import adr.spine.pure.PerformMode
import adr.spine.pure.StepIndex
import adr.spine.pure.StepRecord
import adr.spine.pure.Timestamp

/** The append-only timeline. append() returns the offset — the origin of every effect key. */
class InMemoryBus : Bus {
    private val recorded = mutableListOf<StepRecord>()

    override fun append(record: StepRecord): StepIndex {
        val index = StepIndex(recorded.size)
        recorded += record
        return index
    }

    override fun records(): List<StepRecord> = recorded.toList()
}

/** A clock that never moves — for tests that care about a value, not about ordering. */
fun fixedClock(at: Long): Clock = object : Clock {
    override fun now(): Timestamp = Timestamp(at)
}

/** A clock that advances by `step` on every read. First read returns start + step. */
fun movingClock(start: Long, step: Long): Clock = object : Clock {
    private var t = start
    override fun now(): Timestamp {
        t += step
        return Timestamp(t)
    }
}

fun sequentialIds(prefix: String = "c"): IdSource = object : IdSource {
    private var n = 0
    override fun next(): CommandId = CommandId("$prefix${++n}")
}

/**
 * The audit tap: records every KeyedEffect that crossed the seam, then forwards it.
 * Keys and timestamps are both recorded, so a golden effect sequence compares them.
 */
class RecordingSink(private val downstream: Sink? = null) : Sink {
    val performed = mutableListOf<KeyedEffect>()

    override fun perform(keyed: KeyedEffect, mode: PerformMode) {
        performed += keyed
        downstream?.perform(keyed, mode)
    }
}

/**
 * The recovery-path sink (14.6): the same KeyedEffect delivered twice fires ONCE,
 * because the key was derived from the committed step index and is therefore stable
 * across a crash and a restart.
 */
class DedupingSink(private val downstream: Sink? = null) : Sink {
    private val seen = mutableSetOf<EffectKey>()
    val fired = mutableListOf<Effect>()

    override fun perform(keyed: KeyedEffect, mode: PerformMode) {
        if (!seen.add(keyed.key)) return
        fired += keyed.effect
        downstream?.perform(keyed, mode)
    }
}

File: examples/typescript/src/spine/boundary/in-memory.ts
// ── spine/boundary/in-memory — the fakes the boundary is wired with ────────
// This is the ONE file in the system allowed module-level mutable state
// (check C10), and even here it is confined to closures. Everything an
// application would bind to a real client is a port; these are the offline
// bindings the demo and the tests use.

import type { EffectBase } from "../pure/effect";
import type { StepIndex, Timestamp, CommandId } from "../pure/ids";
import type { KeyedEffect } from "../pure/keyed-effect";
import { keyOf } from "../pure/keyed-effect";
import type { StepRecord } from "../pure/step-record";
import type { Bus } from "../ports/bus";
import type { Clock } from "../ports/clock";
import type { IdSource } from "../ports/id-source";
import type { PerformMode, Sink } from "../ports/sink";

export class InMemoryBus implements Bus {
  private readonly log: StepRecord[] = [];

  append(record: StepRecord): StepIndex {
    this.log.push(record);
    return this.log.length - 1;
  }

  records(): readonly StepRecord[] {
    return this.log;
  }
}

export function fixedClock(at: Timestamp): Clock {
  return { now: () => at };
}

/** A MOVING clock. Every replay test uses one: a frozen clock cannot tell a
 *  faithful re-fold from a lucky one (F8). */
export function movingClock(start: Timestamp, step: Timestamp): Clock {
  let t = start - step;
  return {
    now: () => {
      t += step;
      return t;
    },
  };
}

export function sequentialIds(prefix = "c"): IdSource {
  let n = 0;
  return {
    next: (): CommandId => {
      n += 1;
      return `${prefix}${n}`;
    },
  };
}

/** Records every descriptor that crosses the perform seam — keys, timestamps
 *  and all — then delegates. The recording is what a golden effect sequence is
 *  compared against; the delegate is what actually touches the world. */
export class RecordingSink implements Sink {
  readonly performed: KeyedEffect<EffectBase>[] = [];

  constructor(private readonly inner?: Sink) {}

  perform(keyed: KeyedEffect<EffectBase>, mode: PerformMode): void {
    this.performed.push(keyed);
    this.inner?.perform(keyed, mode);
  }
}

/** The RECOVERY sink (14.6): re-driving a committed timeline after a crash must
 *  not fire an irreversible effect twice. It dedupes on `KeyedEffect.key` — the
 *  key the boundary derived from the committed step index, which is why the
 *  same confirm re-driven twice pages on-call exactly once. */
export class DedupingSink implements Sink {
  readonly fired: KeyedEffect<EffectBase>[] = [];
  private readonly seen = new Set<string>();

  constructor(private readonly inner?: Sink) {}

  perform(keyed: KeyedEffect<EffectBase>, mode: PerformMode): void {
    const key = keyOf(keyed.key);
    if (this.seen.has(key)) return;
    this.seen.add(key);
    this.fired.push(keyed);
    this.inner?.perform(keyed, mode);
  }
}
